### PR TITLE
Added dxFlushCancelledOrders command/API

### PR DESCRIPTION
**Pull request by amdn**
An optional parameter is minimum age in milliseconds for a cancelled order to be flushed, default is 0 (flush all cancelled orders).

Orders are searched for in m_transactions and m_historicTransactions in xbridge App.
Both are std::map, a call to std::map::erase() is used to remove transactions.  If the reference count of the shared_ptr is one, the destructor should be called for the managed object.

The return value contains an array with the list of flushed orders, including the id, txtime, and the use_count (the reference count of the shared_ptr).

This compiles on MacOS, and has been minimally tested with blocknetdx-cli.

### CODE REVIEW

File: src/xbridge/xbridgeapp.cpp [hunk 1]
```diff
@@ -805,6 +809,33 @@ std::map<uint256, xbridge::TransactionDescrPtr> App::history() const
...
+
+    for(auto mp : maps) {
+        for(auto it = mp->begin(); it != mp->end(); ) {
+            const TransactionDescrPtr & ptr = it->second; 
+            if (ptr->state == xbridge::TransactionDescr::trCancelled
+                && ptr->txtime < keepTime) {
+                list.emplace_back(ptr->id,ptr->txtime,ptr.use_count());
+                mp->erase(it++);
+            } else {
+                it++;
```
Comment (corvaxx) [inline src/xbridge/xbridgeapp.cpp]: > `We're Little-Enders and the enemy are Big-Enders.`

iterator& operator++()
{    // preincrement
    ++(*(const_iterator *)this);
    return (*this);
}

iterator operator++(int)
{    // postincrement
    iterator _Tmp = *this;
    ++*this;
    return (_Tmp);
}

:)))

File: src/xbridge/xbridgeapp.cpp [hunk 2]
```diff
@@ -39,6 +39,10 @@
 
 #include "posixtimeconversion.h"
 
+using TransactionMap    = std::map<uint256, xbridge::TransactionDescrPtr>;
```
Comment (corvaxx) [inline src/xbridge/xbridgeapp.cpp]: It's not a good practice to override the types in this place
```
App
{
    typedef std::map<uint256, TransactionDescrPtr> TransactionMap;
    TransactionMap m_transactions;
}
.
.
.

App::TransactionMap * something;
```
imho this better. but this is minor

File: src/xbridge/xbridgeapp.cpp (hunk 1 above)
Comment (amdn) [reply src/xbridge/xbridgeapp.cpp]: I first thought of using std::remove_if but that is not available for std::map.
There is a stackoverflow answer that suggests using the post-increment on the iterator as argument to std::map::erase(), such that the element erased is the one the iterator initially points to (passed as an rvalue to erase()) and the iterator used in the loop is incremented to point at the next element.  The reason this is necessary is that erase() will invalidate the iterator that is passed as parameter.  See https://stackoverflow.com/questions/800955/remove-if-equivalent-for-stdmap.

Comment (corvaxx) [reply src/xbridge/xbridgeapp.cpp]: yes, line 829 - post-increment
but in line 831 post-increment produces parasitic copy of iterator
thats minor :)

File: src/xbridge/xbridgeapp.cpp (hunk 2 above)
Comment (amdn) [reply src/xbridge/xbridgeapp.cpp]: I copied that line from

File: src/xbridge/xbridgeapp.cpp (hunk 1 above)
Comment (amdn) [reply src/xbridge/xbridgeapp.cpp]: Aha!  I agree with you there... will change that one, thank you.

File: src/xbridge/xbridgeapp.cpp (hunk 2 above)
Comment (corvaxx) [reply src/xbridge/xbridgeapp.cpp]: maybe. or leave as is

**Pull request merged**